feat: Add the ability to skip install of local source dependencies

olivier-lacroix opened the pull request:
Fixes https://github.com/prefix-dev/pixi/issues/1652

This PR implements an additional `--skip` argument for `pixi install`, that will skip installation of packages 

As this argument is intended to be used in cases where these local dependencies may be absent, we do not want to update the lockfile, and the use of `--skip` requires the "--frozen" argument.

With this PR, a dockerfile using pixi may look like

```
FROM python:3.12.2-slim

WORKDIR /app

COPY pixi.lock pixi.toml .

RUN pixi install --frozen --skip package-name

COPY sample-package . 

RUN pixi install --frozen
```

ruben-arts commented on src/cli/install.rs:
```diff
@@ -25,6 +25,10 @@ pub struct Args {
     pub all: bool,
+
+    /// Do not install local path dependencies. Requires --frozen
+    #[arg(long, requires = "frozen")]
```
Why would `--locked` not be correct here?

olivier-lacroix replied:
Because with `--locked`, pixi will need the path dependency when checking the lockfile. which does not work if said dependency is not present.

ruben-arts reviewed (changes requested):
Thanks @olivier-lacroix, would you be able to add a test to validate that this works for both conda and PyPI path dependencies. And add the CLI documentation?

olivier-lacroix commented:
> Thanks @olivier-lacroix, would you be able to add a test to validate that this works for both conda and PyPI path dependencies. And add the CLI documentation?

Can do yes :-)

olivier-lacroix commented:
@ruben-arts what sort of test(s) do you think would be best? unit test(s) and/or integration tests?

ruben-arts commented:
For Pypi you might use the `src/install_pypi/plan/test` code. But for the full picture I would write an integration test. Please try to keep these fast, small packages from the local test data channels.

olivier-lacroix commented on src/lib.rs:
```diff
@@ -7,7 +7,7 @@ pub mod environment;
 mod install_pypi;
 pub mod lock_file;
-mod prefix;
+pub mod prefix;
```
Made this public to be able to test if a package was in the prefix. Is this ok?

olivier-lacroix commented:
@ruben-arts integration test added. Let me know what you think.

olivier-lacroix commented on src/lock_file/update.rs:
```diff
@@ -483,6 +498,10 @@ impl<'p> LockFileDerivedData<'p> {
             .conda_packages(platform)
+            .map(|packages| {
+                packages
+                    .filter(|p| !no_path_dependencies || matches!(p.location(), UrlOrPath::Url(_)))
```
Will this exclude local 'built' conda packages? intent is to exclude source packages.

tdejager replied:
I guess it would be better to do `p.is_source().is_some()`

olivier-lacroix replied:
Done @tdejager . Does something similar exist for Pypi records? or are path pypi record always source ?

What do you think of the naming of the flag? I did not want to use 'source' there, as some remote pypi packages can be source packages...

tdejager replied:
No, unfortunately not yet, and a path could be a built dist. I'd guess you would need to check the extension if it is a wheel file..

tdejager commented:
Yeah you are right that should be fine :) only skipping the folders then, I
suppose. You would still want to install sdists from paths and direct urls
right? Or do you want to skip those as well with the flag set?

On Tue, 18 Feb 2025 at 22:30, Olivier Lacroix ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/lock_file/update.rs
> <https://github.com/prefix-dev/pixi/pull/3092#discussion_r1960619329>:
>
> > @@ -483,6 +498,10 @@ impl<'p> LockFileDerivedData<'p> {
>
>          Ok(locked_env
>              .conda_packages(platform)
> +            .map(|packages| {
> +                packages
> +                    .filter(|p| !no_path_dependencies || matches!(p.location(), UrlOrPath::Url(_)))
>
> @tdejager <https://github.com/tdejager> would detecting if the path is a
> folder or not work?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prefix-dev/pixi/pull/3092#discussion_r1960619329>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADF4XRTROOCA7B45NFCVQD2QORADAVCNFSM6AAAAABWYRQH5OVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMRVGA3TMMZXGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

olivier-lacroix commented:
> Yeah you are right that should be fine :) only skipping the folders then, I suppose. You would still want to install sdists from paths and direct urls right? Or do you want to skip those as well with the flag set?

I guess so. the intent is really to skip packages that are being developed.

Quick one, is there a straightforward way to know if a `Utf8TypedPathBuf` is a directory? not too sure how to convert this to a type that can tell me if its a directory, without going through strings, which, I assume, could be pretty brittle.

olivier-lacroix commented:
> I think `as_std_path().is_dir()`, does that work? 

That'd be nice, but this is available in the `camino` crate, not the `typed-path` crate that is in use here . I have based the logic on file extension: A location is deemed a 'local source' if it is a path that is NOT a wheel or an egg file

olivier-lacroix commented:
Alright, @tdejager @ruben-arts this PR is about ready to be merged I think.

@tdejager , a `--skip` flag could definitely be added if there is demand for it. Not sure what use cases there would be for ignoring a specific package though. I propose to defer adding this until some users ask for it?

I was not 100% happy with the flag name `--no-path-dependencies`, as local built artifact are _not_ skipped, and switched to `--skip-local-sources`. this would make it consistent with a potential future `--skip` flag as well 😃 . Happy with that?

tdejager commented:
Hey @olivier-lacroix, @ruben-arts and myself have been discussing this PR offline just now, and we wanted to report back. So while we do think the idea is good there are a couple of things, also because of my suggestion that do not work out. So supposing you still want to skip development dependencies:

1. I wrongly suggested `as_source` as that also includes git dependencies that you do want to include in the initial layering.
2. I think you also still want to include the sdists that would not be excluded, e.g: `.zip`, `.tar.gz` and maybe other extensions as well: https://github.com/astral-sh/uv/blob/f9c1684b96d1b3a88c4d1f4adb4db5e018a7d54f/crates/uv-distribution-filename/src/extension.rs#L29

Additionally, @ruben-arts and I discussed that we think that a general `--skip` makes more sense, as it would simplify some of the logic and you would need to be very explicit what you skip.

We are really sorry about the tardiness regarding this PR, and totally understand that it would be to much work to change this so happy to take this over from you if you do not want to make the change.

olivier-lacroix commented:
@tdejager fair enough. I'll implement it.

olivier-lacroix commented:
@tdejager finally got back to this 😄 ! `pixi install` and `pixi reinstall` now accept a `--skip` flag, allowing to explicitly skip packages.

olivier-lacroix commented on src/cli/reinstall.rs:
```diff
@@ -41,6 +41,11 @@ pub struct Args {
+    /// Skip installation of specific packages present in the lockfile. Requires --frozen
+    /// This can be useful in particular for skipping local source dependencies.
+    #[arg(long, requires = "frozen")]
+    pub skip: Option<Vec<String>>,
```
Not 100% how `--skip` should / will interact with `--packages` here. any thought?

olivier-lacroix commented on tests/integration_rust/install_tests.rs:
```diff
@@ -322,6 +323,63 @@ fn create_uv_environment(prefix: &Path, cache: &uv_cache::Cache) -> PythonEnviro
+/// Test `pixi install --frozen --skip` functionality
+#[tokio::test]
+#[cfg_attr(not(feature = "slow_integration_tests"), ignore)]
+async fn install_frozen_skip() {
```
I don't think so @tdejager as I need to test if the package is / is not added to the prefix. 

Ideally, this could be a recipe / manifest that pixi builds & installs. I am not sure what test data would work for that purpose.

olivier-lacroix replied:
Okay thanks! I have used this one for now. it is indeed not the fastest ...

olivier-lacroix commented:
Alright, I

- removed the `--skip` for reinstall, this can be added later if needed.
- added the conda test

this PR should be good to go I think @tdejager

tdejager reviewed (commented):
Nice! User tested it a bit and it all seems to work. One final thing I would like to see, if thats possible is that we give a warning and do not change the installation method if a non-existing name has been added.

Now currently if we make a typo - z in click*z*:

```
 pixi i --frozen --skip clickz                                                                                                                                                                                                    
✔ The default environment has been installed excluding clickz.
```

Would be better if a tracing warn was printed and the excluding is removed.

Thanks!

olivier-lacroix commented:
@tdejager hum, this isn't as straightforward as I thought 😓. My plan was to bubble up the call stack the packages that were actually skipped and report on that; but, if/when the cache is hit, we can't know this... any idea?

olivier-lacroix commented:
Alright, hopefully getting there @tdejager 😅 .

I have:

- Implemented a public method `get_skipped_package_names` that can test which packages in the lock file will be matched against a list of names, covering the cache hit case. Thanks for the suggestion!
- Moved the filtering logic to the `LockedPackageRef` level (i.e. before specification into pypi or conda packages/records), which simplifies things a fair bit (I think).
- Added a test and modified stdout / stderr output to cover your typo case.

Let me know what you think!

tdejager commented:
@pavelzw might be worth it to update the docker documentation, this might be useful for setting up correct layering.

olivier-lacroix commented:
@ruben-arts this PR requires your approval to be merged.

ruben-arts reviewed (approved):
Thanks @olivier-lacroix !
